firedrake/__init__.py

```diff
@@ -100,3 +100,31 @@
 from firedrake._version import get_versions
 __version__ = get_versions()['version']
 del get_versions
+
+# Try to detect threading and either disable or warn user
+# Threading may come from
+# - OMP_NUM_THREADS: openmp,
+# - OPENBLAS_NUM_THREADS: openblas,
+# - MKL_NUM_THREADS: mkl,
+# - VECLIB_MAXIMUM_THREADS: accelerate,
+# - NUMEXPR_NUM_THREADS: numexpr
+# We only handle the first two cases
+from ctypes import cdll
+from ctypes.util import find_library
+try:
+    _openblas_lib = find_library('openblas')
+    _openblas_dll = cdll.LoadLibrary(_openblas_lib)
+    _openblas_dll.openblas_set_num_threads(1)
+except (OSError, AttributeError):
+    warning('Cannot set OpenBLAS threads, if you are using another BLAS'
+            'implementation, be sure to limit the number of threads to 1')
+
+# OMP_NUM_THREADS can be set to a comma-separated list of positive integers
+try:
+    _omp_num_threads = int(os.environ.get('OMP_NUM_THREADS'))
+except (ValueError, TypeError):
+    _omp_num_threads = None
+if (_omp_num_threads is None) or (_omp_num_threads > 1):
+    warning('OMP_NUM_THREADS is not set or is set to a value greater than 1,'
+            ' we suggest setting OMP_NUM_THREADS=1 to improve performance')
+del _openblas_lib, _openblas_dll, _omp_num_threads, os, cdll, find_library
```

Comment:
Wrap in functions?

Reply:
Not sure if that gains us anything, we don't actually want users calling these functions. Will change if turning this into functions is the preferred style, otherwise I'll leave it

Reply:
I just thought you could make functions and `del function_name` afterwards (don't need to delete so many dels), minor nitpick though.

Reply:
OIC, True, more lines of code though! :laughing: